include/drivers/regulator.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2019-2020 Peter Bigot Consulting, LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief API for regulators controlled by a GPIO.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_GPIO_REGULATOR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_GPIO_REGULATOR_H_
+
+#include <zephyr/types.h>
+#include <drivers/gpio.h>
+#include <sys/onoff.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Driver-specific API functions to support regulator control.
+ */
+struct regulator_driver_api {
+	int (*enable)(struct device *dev, struct onoff_client *cli);
+	int (*disable)(struct device *dev);
+};
+
+/**
+ * @brief Enable a regulator.
+ *
+ * Reference-counted request that a regulator be turned on.  This is
+ * an asynchronous operation; if successfully initiated the result
+ * will b e communicated through the @p cli parameter.
+ *
```

Comment:
so caller has to spin for the result like the init function `regulator_gpio_init` does?

Reply:
No; notification can be spin-wait, callback, or semaphore; everything supported by [sys_notify](https://docs.zephyrproject.org/latest/reference/misc/notify.html#async-notification).  Caller's choice.